### PR TITLE
Add presentedOfferingContext support to custom paywall impression events

### PR DESCRIPTION
**Pull request by rickvdl**
Based on PHC PR https://github.com/RevenueCat/purchases-hybrid-common/pull/1665.

## API

We will now send the `presentedOfferingContext` from the passed `Offering` along with the event. This PR adds the new API variant that takes an `Offering` instead of a plain string `offeringId`. The latter is now deprecated in favor of this new API variant. We will derive the `presentedOfferingContext` from this `Offering` and send it along with the request.

```
await Purchases.trackCustomPaywallImpression({ paywallId: "", offering })
```

---

> [!NOTE]
> **Medium Risk**
> Changes analytics event shape and bridge null-handling for paywall impressions; backward compatible via deprecated `offeringId` but incorrect offering/package data could skew placement analytics.
> 
> **Overview**
> **Custom paywall impression tracking** now accepts a `PurchasesOffering` via `trackCustomPaywallImpression({ offering, paywallId })`. The JS layer derives `offeringId` and `presentedOfferingContext` from the offering’s first available package (placement/targeting), prefers that over legacy `offeringId`, and always forwards a normalized native payload with explicit `null` fields.
> 
> **Native bridges** strip nulls before hybrid common on Android (`mapWithoutNullValues`, also reused for `setAppstackAttributionParams`) and map `NSNull` to omitted keys on iOS for this call. Tests, API testers, Jest mocks, web stub, and the purchase tester **Custom Paywall** screen were updated to exercise offering-based tracking and show the last tracked event.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 14227268e64d79467fed77c7d625e5d9bea2c4a9. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/purchases.ts [hunk 1]
```diff
@@ -1976,14 +1990,40 @@ export default class Purchases {
...
-    params?: TrackCustomPaywallImpressionOptions
+    params?: TrackCustomPaywallImpressionOptions | null
   ): Promise<void> {
     await throwIfNotConfigured();
-    RNPurchases.trackCustomPaywallImpression(params ?? {});
+
+    const presentedOfferingContext =
+      params?.presentedOfferingContext ??
+      params?.offering?.availablePackages?.[0]?.presentedOfferingContext;
+    const offeringId =
```
Comment (rickvdl) [inline src/purchases.ts:2054]: Since the `offeringId` parameter is now deprecated (and will eventually be removed) I figured it made more sense consider the `Offering` the source of truth if passed, and we don't consider the passed `offeringId` as an 'override'.

File: src/purchases.ts [hunk 2]
```diff
@@ -1976,14 +1986,37 @@ export default class Purchases {
    *
    * @param params - Optional parameters for the impression event.
    * @param params.paywallId - Optional identifier for the custom paywall being shown.
+   * @param params.offering - Optional offering associated with the custom paywall.
    * @param params.offeringId - Optional identifier for the offering associated with the custom paywall.
    *   If not provided, the SDK will use the current offering identifier from the cache.
    */
   public static async trackCustomPaywallImpression(
-    params?: TrackCustomPaywallImpressionOptions
+    params?: TrackCustomPaywallImpressionOptions | null
```
Comment (tonidero) [inline src/purchases.ts]: We should add an API test now for null value if we want to support this... Also just wondering if we do want to do this? Seems like keeping it as before is enough?

File: src/purchases.ts [hunk 3]
```diff
@@ -1976,14 +1986,37 @@ export default class Purchases {
...
+    const offeringId =
+      params?.offering?.identifier ??
+      params?.offeringId;
+    const eventData: {
+      paywallId?: string | null;
+      offeringId?: string | null;
+      presentedOfferingContext?: PresentedOfferingContext;
+    } = {};
+
+    if (params?.paywallId !== undefined) {
```
Comment (tonidero) [inline src/purchases.ts]: This is only checking for undefined, but I think it could also be null? I think we can just do
```suggestion
    if (params?.paywallId) {
```

File: src/purchases.ts [hunk 4]
```diff
@@ -1976,14 +1986,37 @@ export default class Purchases {
...
+      eventData.paywallId = params.paywallId;
+    }
+    if (offeringId !== undefined) {
```
Comment (tonidero) [inline src/purchases.ts]: I think this could also be null, so same as above.

File: src/purchases.ts [hunk 5]
```diff
@@ -1976,14 +1986,37 @@ export default class Purchases {
...
+      eventData.offeringId = offeringId;
+    }
+    if (presentedOfferingContext !== undefined && presentedOfferingContext !== null) {
```
Comment (tonidero) [inline src/purchases.ts]: Could maybe be simplified I think for a "truthy" check
```suggestion
    if (presentedOfferingContext) {
```

**Review by tonidero — COMMENTED** [review]
A couple of small things but looking good!

File: src/purchases.ts (hunk 4 above)
Comment (rickvdl) [reply src/purchases.ts]: Good point, I've simplified all of this now :)

**Review by tonidero — APPROVED** [review]
I think this makes sense!!

File: android/src/main/java/com/revenuecat/purchases/react/RNPurchasesModule.java [hunk 6]
```diff
@@ -741,4 +735,54 @@ private static String replacementModeFromProrationMode(int prorationMode) {
...
+
+    private static HashMap<String, Object> toNonNullHashMap(ReadableMap data) {
+        HashMap<String, Object> dataMap = new HashMap<>();
+        for (Map.Entry<String, Object> entry : data.toHashMap().entrySet()) {
+            if (entry.getValue() != null) {
+                dataMap.put(entry.getKey(), entry.getValue());
+            }
+        }
+        return dataMap;
+    }
```
Comment (ajpallares) [inline android/src/main/java/com/revenuecat/purchases/react/RNPurchasesModule.java]: Why not replace this with a just calling `mapWithoutNullValues`? `mapWithoutNullValues` seems more thorough.

File: android/src/main/java/com/revenuecat/purchases/react/RNPurchasesModule.java [hunk 7]
```diff
@@ -741,4 +735,54 @@ private static String replacementModeFromProrationMode(int prorationMode) {
...
+        if (value instanceof Map) {
+            HashMap<String, Object> dataMap = new HashMap<>();
+            Map<?, ?> originalMap = (Map<?, ?>) value;
+            for (Map.Entry<?, ?> entry : originalMap.entrySet()) {
+                Object filteredValue = valueWithoutNullValues(entry.getValue());
+                if (entry.getKey() instanceof String && filteredValue != null) {
+                    dataMap.put((String) entry.getKey(), filteredValue);
+                }
+            }
+            return dataMap;
```
Comment (ajpallares) [inline android/src/main/java/com/revenuecat/purchases/react/RNPurchasesModule.java]: Feels to me like we're repeating this logic in several places. Shouldn't we unify it if we're always returning a Map<String, Object>?

**Review by ajpallares — COMMENTED** [review]
Looks good! I have some comments regarding some repeated code

File: android/src/main/java/com/revenuecat/purchases/react/RNPurchasesModule.java (hunk 7 above)
Comment (ajpallares) [reply android/src/main/java/com/revenuecat/purchases/react/RNPurchasesModule.java]: Basically the same comments as I had in https://github.com/RevenueCat/purchases-flutter/pull/1781#discussion_r3419045149

File: android/src/main/java/com/revenuecat/purchases/react/RNPurchasesModule.java (hunk 6 above)
Comment (rickvdl) [reply android/src/main/java/com/revenuecat/purchases/react/RNPurchasesModule.java]: There's a subtle behavior difference, the implementation used by the appstack method only filtered non-null top level items, whereas the new one did so recursively. However now that I think about it again that's probably fine (or even an improvement), so will consolidate these

Comment (ajpallares) [reply android/src/main/java/com/revenuecat/purchases/react/RNPurchasesModule.java]: That's right. I did realize about it, but my intuition told me the same: it looks like an improvement to me

File: android/src/main/java/com/revenuecat/purchases/react/RNPurchasesModule.java (hunk 7 above)
Comment (rickvdl) [reply android/src/main/java/com/revenuecat/purchases/react/RNPurchasesModule.java]: True, now that I've consolidated these it's a lot simpler I think :)

**Review by ajpallares — APPROVED** [review]
Thanks for the iteration on this! It looks cleaner indeed! Good job!
